### PR TITLE
feat: added hardhat_setNextBlockBaseFeePerGas

### PR DESCRIPTION
**Pull request by vbar**
# What :computer: 
Implemented the `hardhat_setNextBlockBaseFeePerGas` API.

# Why :hand:
Requested in https://github.com/matter-labs/era-test-node/issues/150 .

# Evidence :camera:
Added an e2e test, which passes:
```
  hardhat_setNextBlockBaseFeePerGas
    ✔ should allow setting the next block's base fee per gas
```

### CODE REVIEW

File: SUPPORTED_APIS.md [hunk 1]
```diff
@@ -105,7 +105,7 @@ The `status` options are:
 | `HARDHAT` | `hardhat_setCoinbase` | `NOT IMPLEMENTED` | Sets the coinbase address |
 | `HARDHAT` | `hardhat_setLoggingEnabled` | `NOT IMPLEMENTED` | Enables or disables logging in Hardhat Network |
 | `HARDHAT` | `hardhat_setMinGasPrice` | `NOT IMPLEMENTED` | Sets the minimum gas price |
-| `HARDHAT` | `hardhat_setNextBlockBaseFeePerGas` | `NOT IMPLEMENTED` | Sets the base fee per gas for the next block |
+| [`HARDHAT`](#hardhat-namespace) | [`hardhat_setNextBlockBaseFeePerGas`](#hardhat_setnextblockbasefeepergas) | `SUPPORTED` | Sets the base fee per gas for the next block |
```
Comment (AnastasiiaVashchuk) [inline SUPPORTED_APIS.md:108]: why did you add `(#hardhat_setnextblockbasefeepergas)`?
it is needed only if there is a description here - https://github.com/vbar/era-test-node/blob/set-next-block-base-fee-per-gas/SUPPORTED_APIS.md#network-namespace.

File: e2e-tests/test/hardhat-apis.test.ts [hunk 2]
```diff
@@ -64,6 +64,24 @@ describe("hardhat_mine", function () {
   });
 });
 
+describe("hardhat_setNextBlockBaseFeePerGas", function () {
+  it("should allow setting the next block's base fee per gas", async function () {
+    // Arrange
+    const price = "0x12d687";
```
Comment (AnastasiiaVashchuk) [inline e2e-tests/test/hardhat-apis.test.ts:70]: should we play with types like here? https://github.com/NomicFoundation/hardhat/blob/599a285b37a27cd4a99fd33ef9f4cdad92466af7/packages/hardhat-network-helpers/test/helpers/setNextBlockBaseFeePerGas.ts#L29

File: e2e-tests/test/hardhat-apis.test.ts [hunk 3]
```diff
@@ -64,6 +64,24 @@ describe("hardhat_mine", function () {
...
+
+    // Act
+    const previous = await provider.send("eth_getBlockByNumber", ["latest", false]);
+    await provider.send("hardhat_setNextBlockBaseFeePerGas", [price]);
+    await provider.send("hardhat_mine");
+    const latest = await provider.send("eth_getBlockByNumber", ["latest", false]);
+    await provider.send("hardhat_setNextBlockBaseFeePerGas", [previous.baseFeePerGas]);
```
Comment (AnastasiiaVashchuk) [inline e2e-tests/test/hardhat-apis.test.ts:77]: why do we do this call again?

File: SUPPORTED_APIS.md (hunk 1 above)
Comment (vbar) [reply SUPPORTED_APIS.md]: Just because all the other lines had it - I didn't realize it needs to have a target written as well... I've removed the link.

File: e2e-tests/test/hardhat-apis.test.ts (hunk 2 above)
Comment (vbar) [reply e2e-tests/test/hardhat-apis.test.ts:70]: I'd say no; when testing the server, it doesn't really matter what the types on the client are - the important thing are the communicated JSON strings... Of course, that might be just my personal JS preference...

File: e2e-tests/test/zks-apis.test.ts [hunk 4]
```diff
@@ -30,13 +30,14 @@ describe("zks_estimateFee", function () {
     // Act
     const response: Fee = await provider.send("zks_estimateFee", [transaction]);
     // Assert
-    expect(ethers.BigNumber.from(response.gas_limit)).to.eql(ethers.BigNumber.from("3606743"), "Unexpected gas_limit");
+    expect(ethers.BigNumber.from(response.gas_limit).toNumber()).to.be.within(3606743, 5868728, "Unexpected gas_limit");
```
Comment (AnastasiiaVashchuk) [inline e2e-tests/test/zks-apis.test.ts:33]: what is 5868728?
probably worth to lect comment here.

File: e2e-tests/test/hardhat-apis.test.ts (hunk 3 above)
Comment (vbar) [reply e2e-tests/test/hardhat-apis.test.ts:77]: To leave the test environment as we found it - letting a test change a fee might affect other tests (like the `hardhat_impersonateAccount` case).

File: e2e-tests/test/zks-apis.test.ts [hunk 5]
```diff
@@ -30,13 +30,14 @@ describe("zks_estimateFee", function () {
...
     expect(ethers.BigNumber.from(response.gas_per_pubdata_limit)).to.eql(
       ethers.BigNumber.from("50000"),
       "Unexpected gas_per_pubdata_limit"
     );
-    expect(ethers.BigNumber.from(response.max_fee_per_gas)).to.eql(
-      ethers.BigNumber.from("37500000"),
+    expect(ethers.BigNumber.from(response.max_fee_per_gas).toNumber()).to.be.within(
+      25500297,
+      37500000,
```
Comment (AnastasiiaVashchuk) [inline e2e-tests/test/zks-apis.test.ts:40]: same as above

File: src/namespaces/hardhat.rs [hunk 6]
```diff
@@ -131,4 +131,16 @@ pub trait HardhatNamespaceT {
     /// A `BoxFuture` containing a `Result` with a `bool` representing the success of the operation.
     #[rpc(name = "hardhat_setStorageAt")]
     fn set_storage_at(&self, address: Address, slot: U256, value: U256) -> RpcResult<bool>;
+
+    /// Directly modifies the L1 gas price.
```
Comment (AnastasiiaVashchuk) [inline src/namespaces/hardhat.rs:135]: L1 gas price>

File: e2e-tests/test/zks-apis.test.ts (hunk 4 above)
Comment (vbar) [reply e2e-tests/test/zks-apis.test.ts:35]: OK, commented

Comment (AnastasiiaVashchuk) [reply e2e-tests/test/zks-apis.test.ts:35]: but why exactly 5868728?

File: src/namespaces/hardhat.rs (hunk 6 above)
Comment (vbar) [reply src/namespaces/hardhat.rs:135]: Yes, as stored in the node's fee model - do you want another comment?

File: e2e-tests/test/zks-apis.test.ts (hunk 4 above)
Comment (AnastasiiaVashchuk) [reply e2e-tests/test/zks-apis.test.ts:35]: this approach is not recommendedd. We should not have such situations

Comment (vbar) [reply e2e-tests/test/zks-apis.test.ts:35]: well, I don't know - no more than the original 3606743...

Comment (MexicanAce) [reply e2e-tests/test/zks-apis.test.ts:35]: The reason we have exact values is to ensure that _OUR_ code does not change the estimated gas cost unexpectedly (since it's heavily intertwined in business logic).

This value should only change when updating downstream dependencies and/or updating gas estimation logic on purpose

Comment (vbar) [reply e2e-tests/test/zks-apis.test.ts:35]: Well then, the test found a problem...

File: SUPPORTED_APIS.md (hunk 1 above)
Comment (MexicanAce) [reply SUPPORTED_APIS.md]: We need to add this back with a section dedicated to explaining what this endpoint is and an example curl request

Comment (vbar) [reply SUPPORTED_APIS.md]: The reason I didn't do that is that those sections start with a source link, and I wasn't sure where that would go - just https://github.com/vbar/era-test-node/blob/main/src/node/hardhat.rs ?

Comment (MexicanAce) [reply SUPPORTED_APIS.md]: Yes, wherever the logic is.

Comment (vbar) [reply SUPPORTED_APIS.md:108]: OK, added section.
